scipy/fft/tests/test_helper.py

```diff
@@ -391,55 +391,63 @@ def test_uneven_dims(self, xp):
+        x2 = xp.asarray([0, 1, 2, 3, 4, -5, -4, -3, -2, -1], dtype=xp.float64)
+
+        y = 9 * fft.fftfreq(9, xp=xp_test)
+        assert array_namespace(y) is xp_test
+        y = xp_test.astype(y, xp.float64)
         xp_assert_close(y, x)
-        y = xp.asarray(9 * xp.pi * fft.fftfreq(9, xp.pi, xp=xp), dtype=xp.float64)
+
+        y = xp_test.astype(9 * xp.pi * fft.fftfreq(9, xp.pi, xp=xp_test), xp.float64)
         xp_assert_close(y, x)
 
-        y = xp.asarray(10 * fft.fftfreq(10, xp=xp), dtype=xp.float64)
+        y = xp_test.astype(10 * fft.fftfreq(10, xp=xp_test), xp.float64)
         xp_assert_close(y, x2)
-        y = xp.asarray(10 * xp.pi * fft.fftfreq(10, xp.pi, xp=xp), dtype=xp.float64)
+
+        y = xp_test.astype(10 * xp.pi * fft.fftfreq(10, xp.pi, xp=xp_test), xp.float64)
         xp_assert_close(y, x2)
 
+    def test_device(self, xp):
+        xp_test = array_namespace(xp.empty(0))
+        devices = get_xp_devices(xp)
+        for d in devices:
+            y = fft.fftfreq(9, xp=xp_test, device=d)
+            x = xp_test.empty(0, device=d)
+            assert device(y) == device(x)
+
 
-@skip_xp_backends('array_api_strict', 'cupy',
-                   reasons=['fft not yet implemented by array-api-strict',
-                            'cupy.fft not yet implemented by array-api-compat'])
 class TestRFFTFreq:
 
     def test_definition(self, xp):
-        device = SCIPY_DEVICE
-        try:
-            x = xp.asarray([0, 1, 2, 3, 4], dtype=xp.float64, device=device)
-            x2 = xp.asarray([0, 1, 2, 3, 4, 5], dtype=xp.float64, device=device)
-        except TypeError:
-            # work around the `device` keyword not being implemented in numpy yet
-            x = xp.asarray([0, 1, 2, 3, 4], dtype=xp.float64)
-            x2 = xp.asarray([0, 1, 2, 3, 4, 5], dtype=xp.float64)
-
-        y = xp.asarray(9 * fft.rfftfreq(9, xp=xp), dtype=xp.float64)
+        xp_test = array_namespace(xp.empty(0))
+
+        x = xp.asarray([0, 1, 2, 3, 4], dtype=xp.float64)
+        x2 = xp.asarray([0, 1, 2, 3, 4, 5], dtype=xp.float64)
+
+        y = 9 * fft.rfftfreq(9, xp=xp_test)
+        assert array_namespace(y) is xp_test
+        y = xp_test.astype(y, xp.float64)
         xp_assert_close(y, x)
-        y = xp.asarray(9 * xp.pi * fft.rfftfreq(9, xp.pi, xp=xp), dtype=xp.float64)
+
+        y = xp_test.astype(9 * xp.pi * fft.rfftfreq(9, xp.pi, xp=xp_test), xp.float64)
         xp_assert_close(y, x)
 
-        y = xp.asarray(10 * fft.rfftfreq(10, xp=xp), dtype=xp.float64)
+        y = xp_test.astype(10 * fft.rfftfreq(10, xp=xp_test), xp.float64)
         xp_assert_close(y, x2)
-        y = xp.asarray(10 * xp.pi * fft.rfftfreq(10, xp.pi, xp=xp), dtype=xp.float64)
+
+        y = xp_test.astype(10 * xp.pi * fft.rfftfreq(10, xp.pi, xp=xp_test), xp.float64)
         xp_assert_close(y, x2)
+
+    def test_device(self, xp):
+        xp_test = array_namespace(xp.empty(0))
+        devices = get_xp_devices(xp)
```

Comment:
this does give me `['cpu', 'cuda:0', 'cuda:1']` on a 2 GPU machine with `pytorch`, though CuPy produces `[None]` for the reasons already noted.

Reply:
Evgeni responded from the CuPy end, so I've implemented `is_cupy` now, let me know if it behaves as expected. Otherwise, this should be ready to review. No news on when CuPy is implementing https://github.com/numpy/numpy/pull/25233 as far as I've heard, so test can stay skipped for now.

Reply:
Works as expected for me.